typed-function.js

```diff
@@ -965,21 +965,102 @@
           : restParam2 ? (len1 >= len2)  : (len1 === len2)
     }
 
+    function shallowClone (object) {
+      // we cannot use destructuring {...object} because we still want to support IE11
+      var clone = {}
+
+      Object.keys(object).forEach(key => {
+        clone[key] = object[key];
+      })
+
+      return clone
+    }
+
+    /**
+     * Helper function for `resolveReferences` to throw a meaningful error when
+     * there is an issue with a resolved reference
+     * @param {function | undefined} resolvedReference
+     * @param {string} reference
+     * @param {string} signature
+     */
+    function verifyResolvedReference(resolvedReference, reference, signature) {
+      if (!resolvedReference) {
+        throw new TypeError('Cannot refer to signature "' + signature + '": ' +
+          'reference signature "' + reference + '" not found');
+      }
+
+      if (
+        typeof resolvedReference !== 'function' && (
+          isReferTo(resolvedReference) ||
+          isReferToSelf(resolvedReference)
+        )
+      ) {
+        throw new TypeError('Cannot refer to signature "' + signature + '": ' +
+          'signature is referring to a signature "' + reference + '" which is not yet resolved');
+      }
+    }
+
+    function resolveReferences(rawSignaturesMap, self) {
+      // we must not make changes in the original signatures map
+      var signaturesMap = shallowClone(rawSignaturesMap);
+
+      // resolve references, providing "self"
+      for (var signature in signaturesMap) {
+        if (signaturesMap.hasOwnProperty(signature)) {
+          var fn = signaturesMap[signature];
+
+          if (isReferTo(fn)) {
+            var resolvedReferences = fn.referTo.references.map(reference => {
+              var resolvedReference = signaturesMap[reference];
+
+              verifyResolvedReference(resolvedReference, reference, signature);
+
+              return resolvedReference;
+            })
+            var resolved = fn.referTo.callback.apply(this, resolvedReferences);
+
+            // we need to keep the .referTo object (with references and callback)
+            // to be able to merge typed-functions later on: then we need to
+            // resolve the new signatures again.
+            resolved.referTo = fn.referTo;
+
```

Comment:
While intuitively the comment here before line 1025 is correct, if you comment out line 1025 all of the unit tests succeed. But that's no so surprising, because as far as I could see there are no unit tests for merging functions that have "refer-to"s in them. It seems to me that there certainly should be, so that's one recommended action item here.

But further, I was not actually able to create a test that failed even with this line commented out: there does not seem to be any way to override an existing signature of a typed function in a merge, and there is no way for a merge to "capture" a reference from one of the the merged items to another. That's because a single plain function can't be a refer-to, and so refer-to is only relevant to the typed-function arguments; but since all of the typed-function arguments had to have been fully resolved at their creation times, all of their references were satisfied "internally" to each of themselves, and so after the merge all of them will still resolve to exactly the same thing that they would have resolved to on their own.

But note this is **not** a recommendation to remove this code propagating the .referTo object. It **is** (partly) an argument to accept PR #142, because then one could merge new additional signatures with an existing typed-function that would referTo signatures that were on that existing typed-function, greatly increasing the power to leverage existing behavior when extending.

But even with #142, you're only resolving new "referTo"s, not re-resolving previously resolved "referTo"s  and arriving at a new outcome. That's because of the prohibition on overriding an existing signature. However, I _still_ think the referTo property should be kept, because I think that sooner or later typed-functions _should_ have the ability to modify a previously existing signature by some reasonable API, and for that purpose it would be best to have been keeping the referTo around.

 (But all that said if you wanted to drop the referTo for now, I racked my brains but could not come up with a way under the current API that it could make a difference, or even with #142 present make a difference. If you have such an example, it should _definitely_ be captured in a unit test, I should think.)

Reply:
Hm, yes, a test for that is missing. You may need it when merging typed-functions. It is a bit of an edge case, you cannot produce an issue with this with the public API of creating and merging two functions directly. It can occur when you manually pick signatures from two existing typed-functions, and put them in one new typed-function.

(I had created a test for the same mechanism `resolvedSelf.referToSelf = fn.referToSelf` and just checked, the unit test fails when I remove that line of code).

I've now added a unit test for it, it is named 'should be able to use referTo when merging signatures from multiple typed-functions'.

Reply:
Wow, that is quite an edge case, and frankly it feels a bit wrong that the outcome of `typed(tf1, tf2)` is different from `typed({...tf1.signatures, ..tf2.signatures})`. But anyhow, for the point in question, I am definitely not against keeping the .referTo property of the signature, so happy to leave this item be now.

Reply:
> and frankly it feels a bit wrong that the outcome of `typed(tf1, tf2)` is different from `typed({...tf1.signatures, ..tf2.signatures})`.

The reason is just how merging of two objects works: if both objects have the same key, the key of one of the two will be overwritten. So:

```js
const obj1 = { number: fn1 }
const obj2 = { number: fn2 }

// merging objects with the same keys will silently overwrite the first one with the second
const obj3 = { ...obj1, ...obj2 } // obj3 will be { number: fn2 }

const tf1 = typed(obj1)
const tf2 = typed(obj2)
const tf3 = typed(tf1, tf2) // will throw an exception about conflicting (duplicate) signatures
```